### ISSUE TITLE
packer does not support google oslogin

### ISSUE THREAD

**Issue by Tahvok** [hashicorp/packer#7663]
Using packer version: 1.4.1

If you bring up your instances with oslogin enabled packer is unable to connect to them. Even if ssh-agent is present with correct key and correct ssh_username is supplied.
Basically to test this, when packer has created the instance I was able to connect from the same source instance via ssh, but packer still couldn't. I would expect packer to at least try to use the running ssh agent.

Log:
https://gist.github.com/Tahvok/9325ab2961fb0bf9b8e89e70fa3b32a0

Builder to reproduce:
```
"builders": [
        {
            "type": "googlecompute",
            "project_id": "project-id",
            "zone": "us-central1-a",
            "instance_name": "coreos-stable-2079-4-0-v20190515",
            "source_image": "coreos-stable-2079-4-0-v20190515",
            "ssh_username": "service-account-username",
            "image_name": "image-name",
            "metadata": {
              "enable-oslogin": "TRUE"
            }
        }
    ]
```

**Comment by azr** [comment]
Hello @Tahvok, thanks for opening. Weird, it worked for me using :

```json
            "machine_type": "n1-standard-1",
            "source_image": "debian-7-wheezy-v20150127",
            "ssh_username": "packer",
            "type": "googlecompute",
            "zone": "{{ user `zone`}}",
            "metadata": {
              "enable-oslogin": "TRUE"
            }
```

**Comment by azr** [comment]
But now this looks like to me it is a bug with the os / or cloud provider and not packer 🤔

**Comment by Tahvok** [comment]
Have you completely enabled oslogin on google?
https://cloud.google.com/compute/docs/instances/managing-instance-access

I have found issue #6341 where it didn't work for @rosstimson as well, and he had to workaround it by setting enable-oslogin to false in metadata

**Comment by azr** [comment]
I have enabled it just for the individual instance :
> Enable the OS Login feature on your project or on individual instances.

**Comment by rhencke** [comment]
I believe this is a bug/missing feature in the Google Cloud provider.

There are two ways I am aware of for managing SSH key access with GCP:
* [SSH metadata](https://cloud.google.com/compute/docs/instances/adding-removing-ssh-keys) 
* [OS Login](https://cloud.google.com/compute/docs/instances/managing-instance-access)

From what I can tell through reading https://github.com/hashicorp/packer/blob/78c0b7bd9c1c2464c407fadbed3da4e94e627796/builder/googlecompute/step_create_instance.go it seems that the cloud provider currently only handles SSH metadata-based key access, and doesn't appear to support OS Login.

**Comment by rickard-von-essen** [comment]
@rhencke pretty sure everything is there o  the GCP side, someone just need to take some tune and investigate exactly how this should be done and implement the support. I expect it to be fairly straightforward do.

**Comment by junjihashimoto** [comment]
Hi,

Our system uses enable-oslogin with project' metadata.
We can build a image by using following builders-settings.


```
            "metadata": {
              "enable-oslogin": "FALSE"
            }
```

**Comment by thnee** [comment]
Can confirm, this is still a problem. I used to have OS Login disabled, packer was working fine, when enabling it on the project level, packer stopped working with this output.

```
==> googlecompute: Waiting for SSH to become available...
==> googlecompute: Error waiting for SSH: Packer experienced an authentication error when trying to connect via SSH. This can happen if your username/password are wrong. You may want to double-check your credentials as part of your debugging process. original error: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
==> googlecompute: Packer experienced an authentication error when trying to connect via SSH. This can happen if your username/password are wrong. You may want to double-check your credentials as part of your debugging process. original error: ssh: handshake failed: ssh: unable to authenticate, attempted methods [none publickey], no supported methods remain
==> googlecompute: Step "StepConnect" failed
```

Fortunately, the workaround is simple, just add this to the `googlecompute` builder.

```json
  "metadata": {
    "enable-oslogin": "FALSE"
  }
```

**Comment by dvmorris** [comment]
Is there a plan to support this in the future? I just ran into the same issue today.

**Comment by SwampDragons** [comment]
We're not actively working on this, yet, but PRs are always welcome. :)

**Comment by kevinbirch** [comment]
> We're not actively working on this, yet, but PRs are always welcome. :)

Could you roughly estimate the level of effort for someone outside Hashicorp to implement this feature?

**Comment by SwampDragons** [comment]
Well, with one major caveat: I haven't researched how OSlogin works or how to link it to Packer, so that element could be totally simple through the SDK, or downright impossible. But assuming it's doable via the google sdk for go, this shouldn't be too hard to implement. (famous last words). You can get an idea of how the current authentication is handled by tracing the AccountFile through the code: https://github.com/hashicorp/packer/blob/master/builder/googlecompute/config.go#L36 

That gets a little hairy when it calls out to the oauth code, but you don't have to follow it all the way through to see how Packer handles it on input. It's possible oslogin could be simpler; you'd add a feature to that config struct like use_oslogin or something, and call out to the sdk in the Prepare() instead of doing the AccountFile prep in that same Prepare method.

**Comment by cpwc** [comment]
Tested out using `OSLogin` for service accounts, here are some of my findings.

SSH key generated or provided have to be added to gcloud account using this command.
```
gcloud compute os-login ssh-keys add --key-file [path-to-pub-file]
```

Once the command completes it will return some information below. The information can determine the `ssh_username` to be used for communicator.
```
loginProfile:
  name:  [redacted]
  posixAccounts:
  - accountId: [redacted]
    username: [SSH_USERNAME]
  sshPublicKeys:
    [redacted]
```

Depends if you are logging packer as a `service_account` or `user` the `ssh_username` to be used will differ. 

Note: If role given is `Compute OS Admin Login` you will be logged in directly as a user with administrative rights.

Couple of thoughts on how we could potentially implement support for OSLogin.

- Check for specific config to enable the use of `OSLogin`?
- Before/after we create the instance add the SSH key using the command above. There are APIs available for that as well https://cloud.google.com/compute/docs/oslogin/rest
- Replace the `ssh_username` to be used with the one returned from `OSLogin`.
- Proceed as per normal for SSH communicator.

**Comment by cpwc** [comment]
Would like to get some help in obtaining user's email if user is not using `account.json` file and it will fallback to default credentials. As per the API one of the parameter `parent` needs to be of a specific format `users/{user}` which will be user's email address in this case.

https://github.com/googleapis/google-api-go-client/blob/v0.25.0/oslogin/v1/oslogin-gen.go#L680

**Comment by thnee** [comment]
@cpwc When using Ansible, I rely on `gcloud compute os-login describe-profile` to get the needed information about the user.

**Comment by cpwc** [comment]
@thnee yeah. Unfortunately there's no API that could do just that. There's this https://cloud.google.com/compute/docs/oslogin/rest/v1/users/getLoginProfile API but it still require to know the user's email before hand to craft the API request.

**Comment by cpwc** [comment]
👋  hello would love to get some inputs from those who are looking for oslogin support on packer. Would like you guys to test out the binaries from #9469 and see if it's how you would expect oslogin to work based on the work and help from @nywilken

**Comment by eriksw** [comment]
@cpwc To find the email of the default google credentials, see [oauth2/Service.TokenInfo](https://godoc.org/google.golang.org/api/oauth2/v2#Service.Tokeninfo).

Instantiate the Service without configuring credentials in any way, then call that and check the Email in the result struct.

---

For my use case, I would like to have packer use a service account (running within CI) to log into the instance being built. 

My expectation for oslogin support in packer would be for it to "just work" exactly the same as if packer was shelling out to `gcloud compute ssh ... --tunnel-through-iap`. In particular, I do not ever want to have to create, store, or manage the lifecycle of an ssh key—I would expect packer to keep the fact that there is an ssh key involved a fully-ignorable implementation detail just the same as `gcloud compute ssh` does.

**Comment by nywilken** [comment]
@eriksw thanks for pointing out oauth2/Service.TokenInfo. We will take a look. 

> My expectation for oslogin support in packer would be for it to "just work" exactly the same as if packer was shelling out to gcloud compute ssh ... --tunnel-through-iap. In particular, I do not ever want to have to create, store, or manage the lifecycle of an ssh key—I would expect packer to keep the fact that there is an ssh key involved a fully-ignorable implementation detail just the same as gcloud compute ssh does.

For this scenario would you specify the correct username in the case of a service account something like `sa_123456789987654567890987654`? Or would you expect Packer to figure that out as well? 

I think this is what you are alluding to already, but seeing as OSLogin usernames must be specific to an account I want to make sure that dynamically updating the username would be an expected behavior. This is in regards to "Replace the ssh_username to be used with the one returned from OSLogin." mentioned in @cpwc [investigation ](https://github.com/hashicorp/packer/issues/7663#issuecomment-634853946)

An  alternative could be that  Packer errors if the username doesn't match the username provided by OSLogin to avoid surprises.

**Comment by eriksw** [comment]
I would strongly prefer to not have to set `ssh_username` at all when using oslogin.

Having packer correctly determine the username in use the same way `gcloud compute ssh` does would make the same template usable without modifications by both a user who has access to the project and also service account CI.

**Comment by nywilken** [comment]
@eriksw thanks again for the tip on [oauth2/Service.Tokeninfo](https://godoc.org/google.golang.org/api/oauth2/v2#Service.Tokeninfo). I went ahead and added a few changes on top of the awesome work done by @cpwc to handle the dynamic SSH key importing for service accounts. Would you mind giving it a try? 

The test binaries can be found [here](https://app.circleci.com/pipelines/github/hashicorp/packer/5884/workflows/c1ec10f9-a2b9-4582-92a6-5e73710c8542/jobs/63401/artifacts). Documentation for the `use_os_login` option can be found [here](https://github.com/hashicorp/packer/blob/b8fb5b5eace9b39d78e16f1ab97e974568b6e170/website/pages/partials/builder/googlecompute/Config-not-required.mdx)

**Comment by nywilken** [comment]
Hello again, support for OSLogin has been merged and will be available in the v1.6.1 Packer release. If you would like to gain access to this feature before its official release it will be available for use in the next nightly release. There are also test binaries [here](https://app.circleci.com/pipelines/github/hashicorp/packer/5992/workflows/075e806e-b736-48a1-8a1c-42013a638718/jobs/64902/artifacts) if you would like to give the feature a try. Cheers!
